File: app/android/app/src/main/java/com/keyboardn/app/plugins/PurchasesPlugin.kt
package com.keyboardn.app.plugins

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Offerings
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesError
import com.revenuecat.purchases.interfaces.GetOfferingsCallback
import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
import com.revenuecat.purchases.interfaces.PurchaseCallback
import com.revenuecat.purchases.interfaces.ReceiveOfferingsCallback
import com.revenuecat.purchases.interfaces.SyncPurchasesCallback
import com.revenuecat.purchases.models.StoreTransaction
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

@CapacitorPlugin(name = "Purchases")
class PurchasesPlugin : Plugin() {

    private val isoFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    @PluginMethod
    fun configure(call: PluginCall) {
        val apiKey = call.getString("apiKey") ?: run {
            call.reject("Missing apiKey")
            return
        }
        val appUserId = call.getString("appUserId")

        val config = PurchasesConfiguration.Builder(context, apiKey)
            .apply { if (appUserId != null) appUserID(appUserId) }
            .build()

        Purchases.configure(config)

        Purchases.sharedInstance.updatedCustomerInfoListener = { customerInfo ->
            val data = JSObject()
            data.put("customerInfo", convertCustomerInfo(customerInfo))
            notifyListeners("purchaserInfoUpdate", data)
        }

        call.resolve()
    }

    @PluginMethod
    fun getCustomerInfo(call: PluginCall) {
        Purchases.sharedInstance.getCustomerInfo(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                val result = JSObject()
                result.put("customerInfo", convertCustomerInfo(customerInfo))
                call.resolve(result)
            }
            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun getOfferings(call: PluginCall) {
        Purchases.sharedInstance.getOfferings(object : ReceiveOfferingsCallback {
            override fun onReceived(offerings: Offerings) {
                val result = JSObject()
                result.put("offerings", convertOfferings(offerings))
                call.resolve(result)
            }
            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun purchasePackage(call: PluginCall) {
        val identifier = call.getString("identifier") ?: run {
            call.reject("Missing package identifier")
            return
        }
        val offeringIdentifier = call.getString("offeringIdentifier") ?: run {
            call.reject("Missing offering identifier")
            return
        }

        Purchases.sharedInstance.getOfferings(object : ReceiveOfferingsCallback {
            override fun onReceived(offerings: Offerings) {
                val pkg = offerings[offeringIdentifier]?.getPackage(identifier)
                if (pkg == null) {
                    call.reject("Package not found")
                    return
                }

                activity?.let { act ->
                    Purchases.sharedInstance.purchase(act, pkg, object : PurchaseCallback {
                        override fun onCompleted(storeTransaction: StoreTransaction, customerInfo: CustomerInfo) {
                            val result = JSObject()
                            result.put("customerInfo", convertCustomerInfo(customerInfo))
                            result.put("productIdentifier", storeTransaction.productIds.firstOrNull() ?: "")
                            call.resolve(result)
                        }
                        override fun onError(error: PurchasesError, userCancelled: Boolean) {
                            if (userCancelled) {
                                call.reject("User cancelled", "PURCHASE_CANCELLED")
                            } else {
                                call.reject(error.message)
                            }
                        }
                    })
                } ?: call.reject("Activity not available")
            }
            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun restorePurchases(call: PluginCall) {
        Purchases.sharedInstance.restorePurchases(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                val result = JSObject()
                result.put("customerInfo", convertCustomerInfo(customerInfo))
                call.resolve(result)
            }
            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun syncPurchases(call: PluginCall) {
        Purchases.sharedInstance.syncPurchases(object : SyncPurchasesCallback {
            override fun onSuccess(customerInfo: CustomerInfo) {
                call.resolve()
            }
            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun setDebugLogsEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled", false) ?: false
        Purchases.debugLogsEnabled = enabled
        call.resolve()
    }

    // MARK: - Helpers

    private fun convertCustomerInfo(info: CustomerInfo): JSObject {
        val entitlements = JSObject()
        val active = JSObject()
        info.entitlements.active.forEach { (k, v) -> active.put(k, convertEntitlement(v)) }
        val all = JSObject()
        info.entitlements.all.forEach { (k, v) -> all.put(k, convertEntitlement(v)) }
        entitlements.put("active", active)
        entitlements.put("all", all)

        val nonSubs = JSONArray()
        info.nonSubscriptionTransactions.forEach { nonSubs.put(convertNonSubscriptionTransaction(it)) }

        val result = JSObject()
        result.put("entitlements", entitlements)
        result.put("firstSeen", isoFormat.format(info.firstSeen))
        result.put("originalAppUserId", info.originalAppUserId)
        result.put("originalPurchaseDate", info.originalPurchaseDate?.let { isoFormat.format(it) })
        result.put("nonSubscriptionTransactions", nonSubs)
        return result
    }

    private fun convertEntitlement(e: com.revenuecat.purchases.EntitlementInfo): JSObject {
        val result = JSObject()
        result.put("identifier", e.identifier)
        result.put("isActive", e.isActive)
        result.put("willRenew", e.willRenew)
        result.put("latestPurchaseDate", isoFormat.format(e.latestPurchaseDate))
        result.put("originalPurchaseDate", isoFormat.format(e.originalPurchaseDate))
        result.put("expirationDate", e.expirationDate?.let { isoFormat.format(it) })
        result.put("productIdentifier", e.productIdentifier)
        return result
    }

    private fun convertNonSubscriptionTransaction(t: com.revenuecat.purchases.models.Transaction): JSObject {
        val result = JSObject()
        result.put("productIdentifier", t.productIdentifier)
        result.put("purchaseDate", isoFormat.format(t.purchaseDate))
        result.put("transactionIdentifier", t.transactionIdentifier)
        return result
    }

    private fun convertOfferings(offerings: Offerings): JSObject {
        val result = JSObject()
        result.put("current", offerings.current?.let { convertOffering(it) })
        val all = JSObject()
        offerings.all.forEach { (k, v) -> all.put(k, convertOffering(v)) }
        result.put("all", all)
        return result
    }

    private fun convertOffering(offering: Offering): JSObject {
        val pkgs = JSONArray()
        offering.availablePackages.forEach { pkgs.put(convertPackage(it)) }
        val result = JSObject()
        result.put("identifier", offering.identifier)
        result.put("availablePackages", pkgs)
        return result
    }

    private fun convertPackage(pkg: Package): JSObject {
        val result = JSObject()
        result.put("identifier", pkg.identifier)
        result.put("packageType", pkg.packageType.name)
        result.put("offeringIdentifier", pkg.offering)
        result.put("product", convertProduct(pkg.product))
        return result
    }

    private fun convertProduct(product: com.revenuecat.purchases.models.StoreProduct): JSObject {
        val result = JSObject()
        result.put("identifier", product.purchasingData.productIdentifier)
        result.put("description", product.description)
        result.put("title", product.title)
        result.put("price", product.price.amountMicros / 1_000_000.0)
        result.put("priceString", product.price.formatted)
        result.put("currencyCode", product.price.currencyCode)
        return result
    }
}

File: app/ios/App/App/PurchasesPlugin/PurchasesPlugin.swift
import Capacitor
import RevenueCat

@objc(PurchasesPlugin)
public class PurchasesPlugin: CAPPlugin, CAPBridgedPlugin, PurchasesDelegate {
    public let identifier = "PurchasesPlugin"
    public let jsName = "Purchases"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure",           returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCustomerInfo",     returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getOfferings",        returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "purchasePackage",     returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "restorePurchases",    returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncPurchases",       returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setDebugLogsEnabled", returnType: CAPPluginReturnPromise),
    ]

    // PurchasesDelegate - called when customer info updates
    public func purchases(_ purchases: Purchases, receivedUpdated customerInfo: CustomerInfo) {
        notifyListeners("purchaserInfoUpdate", data: ["customerInfo": convertCustomerInfo(customerInfo)])
    }

    @objc func configure(_ call: CAPPluginCall) {
        guard let apiKey = call.getString("apiKey") else {
            call.reject("Missing apiKey"); return
        }
        let appUserId = call.getString("appUserId")
        Purchases.configure(withAPIKey: apiKey, appUserID: appUserId)
        Purchases.shared.delegate = self
        call.resolve()
    }

    @objc func getCustomerInfo(_ call: CAPPluginCall) {
        Purchases.shared.getCustomerInfo { [weak self] customerInfo, error in
            if let error = error { call.reject(error.localizedDescription); return }
            call.resolve(["customerInfo": self?.convertCustomerInfo(customerInfo) ?? [:]])
        }
    }

    @objc func getOfferings(_ call: CAPPluginCall) {
        Purchases.shared.getOfferings { [weak self] offerings, error in
            if let error = error { call.reject(error.localizedDescription); return }
            call.resolve(["offerings": self?.convertOfferings(offerings) ?? [:]])
        }
    }

    @objc func purchasePackage(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier"),
              let offeringIdentifier = call.getString("offeringIdentifier") else {
            call.reject("Missing package identifier or offering identifier"); return
        }
        Purchases.shared.getOfferings { [weak self] offerings, error in
            if let error = error { call.reject(error.localizedDescription); return }
            guard let package = offerings?.offering(identifier: offeringIdentifier)?.package(identifier: identifier) else {
                call.reject("Package not found"); return
            }
            Purchases.shared.purchase(package: package) { [weak self] transaction, customerInfo, error, userCancelled in
                if userCancelled { call.reject("User cancelled", "PURCHASE_CANCELLED"); return }
                if let error = error { call.reject(error.localizedDescription); return }
                call.resolve([
                    "customerInfo": self?.convertCustomerInfo(customerInfo) ?? [:],
                    "productIdentifier": package.storeProduct.productIdentifier
                ])
            }
        }
    }

    @objc func restorePurchases(_ call: CAPPluginCall) {
        Purchases.shared.restorePurchases { [weak self] customerInfo, error in
            if let error = error { call.reject(error.localizedDescription); return }
            call.resolve(["customerInfo": self?.convertCustomerInfo(customerInfo) ?? [:]])
        }
    }

    @objc func syncPurchases(_ call: CAPPluginCall) {
        Purchases.shared.syncPurchases { _, error in
            if let error = error { call.reject(error.localizedDescription); return }
            call.resolve()
        }
    }

    @objc func setDebugLogsEnabled(_ call: CAPPluginCall) {
        Purchases.logLevel = call.getBool("enabled", false) ? .debug : .warn
        call.resolve()
    }

    // MARK: - Helpers

    private func convertCustomerInfo(_ customerInfo: CustomerInfo?) -> [String: Any] {
        guard let customerInfo = customerInfo else { return [:] }
        return [
            "entitlements": [
                "active": customerInfo.entitlements.active.mapValues { convertEntitlement($0) },
                "all": customerInfo.entitlements.all.mapValues { convertEntitlement($0) }
            ],
            "firstSeen": customerInfo.firstSeen.ISO8601Format(),
            "originalAppUserId": customerInfo.originalAppUserId,
            "originalPurchaseDate": customerInfo.originalPurchaseDate?.ISO8601Format() as Any,
            "nonSubscriptionTransactions": customerInfo.nonSubscriptionTransactions.map { convertNonSubscriptionTransaction($0) }
        ]
    }

    private func convertEntitlement(_ e: EntitlementInfo) -> [String: Any] {
        return [
            "identifier": e.identifier,
            "isActive": e.isActive,
            "willRenew": e.willRenew,
            "latestPurchaseDate": e.latestPurchaseDate?.ISO8601Format() as Any,
            "originalPurchaseDate": e.originalPurchaseDate?.ISO8601Format() as Any,
            "expirationDate": e.expirationDate?.ISO8601Format() as Any,
            "productIdentifier": e.productIdentifier
        ]
    }

    private func convertNonSubscriptionTransaction(_ t: StoreTransaction) -> [String: Any] {
        return [
            "productIdentifier": t.productIdentifier,
            "purchaseDate": t.purchaseDate.ISO8601Format(),
            "transactionIdentifier": t.transactionIdentifier
        ]
    }

    private func convertOfferings(_ offerings: Offerings?) -> [String: Any] {
        guard let offerings = offerings else { return [:] }
        return [
            "current": offerings.current.map { convertOffering($0) } as Any,
            "all": offerings.all.mapValues { convertOffering($0) }
        ]
    }

    private func convertOffering(_ offering: Offering) -> [String: Any] {
        return [
            "identifier": offering.identifier,
            "availablePackages": offering.availablePackages.map { convertPackage($0) },
            "metadata": offering.metadata
        ]
    }

    private func convertPackage(_ package: Package) -> [String: Any] {
        return [
            "identifier": package.identifier,
            "packageType": package.packageType.rawValue,
            "offeringIdentifier": package.offeringIdentifier,
            "product": convertProduct(package.storeProduct)
        ]
    }

    private func convertProduct(_ product: StoreProduct) -> [String: Any] {
        return [
            "identifier": product.productIdentifier,
            "description": product.localizedDescription,
            "title": product.localizedTitle,
            "price": product.price,
            "priceString": product.localizedPriceString,
            "currencyCode": product.currencyCode ?? ""
        ]
    }
}
